File: backend/src/main/kotlin/mymoney/domain/model/Goal.kt
package mymoney.domain.model

import kotlinx.datetime.Instant
import java.util.UUID

/**
 * Goal — накопительная цель. currentAmountKopecks меняется отдельными
 * usecase'ами deposit/withdraw, чтобы не путать с прямым редактированием.
 */
data class Goal(
    val id: UUID,
    val familyId: UUID,
    val name: String,
    val targetAmountKopecks: Long,
    val currentAmountKopecks: Long,
    val targetDate: Instant?,
    val createdAt: Instant,
    val updatedAt: Instant,
    val isDeleted: Boolean = false,
) {
    val isCompleted: Boolean get() = currentAmountKopecks >= targetAmountKopecks
    val progressPercent: Int
        get() {
            if (targetAmountKopecks <= 0L) return 0
            val raw = (currentAmountKopecks * 100 / targetAmountKopecks).toInt()
            return raw.coerceIn(0, 100)
        }
}

File: mobile/lib/domain/model/goal.dart
class Goal {
  final String id;
  final String familyId;
  final String name;
  final int targetAmountKopecks;
  final int currentAmountKopecks;
  final DateTime? targetDate;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isDeleted;

  const Goal({
    required this.id,
    required this.familyId,
    required this.name,
    required this.targetAmountKopecks,
    required this.currentAmountKopecks,
    this.targetDate,
    required this.createdAt,
    required this.updatedAt,
    this.isDeleted = false,
  });

  bool get isCompleted => currentAmountKopecks >= targetAmountKopecks;

  int get progressPercent {
    if (targetAmountKopecks <= 0) return 0;
    final raw = (currentAmountKopecks * 100 / targetAmountKopecks).round();
    return raw.clamp(0, 100);
  }

  Goal copyWith({
    String? name,
    int? targetAmountKopecks,
    int? currentAmountKopecks,
    DateTime? targetDate,
    DateTime? updatedAt,
    bool? isDeleted,
    bool clearTargetDate = false,
  }) =>
      Goal(
        id: id,
        familyId: familyId,
        name: name ?? this.name,
        targetAmountKopecks: targetAmountKopecks ?? this.targetAmountKopecks,
        currentAmountKopecks: currentAmountKopecks ?? this.currentAmountKopecks,
        targetDate: clearTargetDate ? null : (targetDate ?? this.targetDate),
        createdAt: createdAt,
        updatedAt: updatedAt ?? this.updatedAt,
        isDeleted: isDeleted ?? this.isDeleted,
      );
}
